flink-runtime/src/main/java/org/apache/flink/runtime/rpc/taskexecutor/TaskExecutor.java

```diff
@@ -35,27 +79,634 @@
+
+			if (memType == MemoryType.HEAP) {
+				long relativeMemSize = (long) (EnvironmentInformation.getSizeOfFreeHeapMemoryWithDefrag() * fraction);
+				if (preAllocateMemory) {
+					LOG.info("Using {} of the currently free heap space for managed heap memory ({} MB)." ,
+						fraction , relativeMemSize >> 20);
+				} else {
+					LOG.info("Limiting managed memory to {} of the currently free heap space ({} MB), " +
+						"memory will be allocated lazily." , fraction , relativeMemSize >> 20);
+				}
+				memorySize = relativeMemSize;
+			} else if (memType == MemoryType.OFF_HEAP) {
+				// The maximum heap memory has been adjusted according to the fraction
+				long maxMemory = EnvironmentInformation.getMaxJvmHeapMemory();
+				long directMemorySize = (long) (maxMemory / (1.0 - fraction) * fraction);
+				if (preAllocateMemory) {
+					LOG.info("Using {} of the maximum memory size for managed off-heap memory ({} MB)." ,
+						fraction, directMemorySize >> 20);
+				} else {
+					LOG.info("Limiting managed memory to {} of the maximum memory size ({} MB)," +
+						" memory will be allocated lazily.", fraction, directMemorySize >> 20);
+				}
+				memorySize = directMemorySize;
+			} else {
+				throw new RuntimeException("No supported memory type detected.");
+			}
+		}
+
+		// now start the memory manager
+		MemoryManager memoryManager;
+		try {
+			memoryManager = new MemoryManager(
+				memorySize,
+				taskManagerConfig.numberOfSlots(),
+				netConfig.networkBufferSize(),
+				memType,
+				preAllocateMemory);
+		} catch (OutOfMemoryError e) {
+			if (memType == MemoryType.HEAP) {
+				throw new Exception("OutOfMemory error (" + e.getMessage() +
+					") while allocating the TaskManager heap memory (" + memorySize + " bytes).", e);
+			} else if (memType == MemoryType.OFF_HEAP) {
+				throw new Exception("OutOfMemory error (" + e.getMessage() +
+					") while allocating the TaskManager off-heap memory (" + memorySize +
+					" bytes).Try increasing the maximum direct memory (-XX:MaxDirectMemorySize)", e);
+			} else {
+				throw e;
+			}
+		}
+
+		// start the I/O manager, it will create some temp directories.
+		IOManager ioManager = new IOManagerAsync(taskManagerConfig.tmpDirPaths());
+
+		if (leaderRetrievalService == null){
+			leaderRetrievalService = LeaderRetrievalUtils.createLeaderRetrievalService(configuration);
+		}
+
+		// create the actor properties (which define the actor constructor parameters)
+		Props tmProps = Props.create(
+			taskManagerClass,
+			taskManagerConfig,
+			resourceID,
+			connectionInfo,
+			memoryManager,
+			ioManager,
+			network,
+			taskManagerConfig.numberOfSlots(),
+			leaderRetrievalService);
+
+		ActorRef taskManagerActorRef;
+		if (taskManagerActorName != null && !taskManagerActorName.equals("")) {
+			taskManagerActorRef = actorSystem.actorOf(tmProps, taskManagerActorName);
+		} else {
+			taskManagerActorRef = actorSystem.actorOf(tmProps);
+		}
+
+		return taskManagerActorRef;
```

Comment:
The `HighAvailabilityService` is just an interface that you implement in a class which you pass to the constructor of the TaskExcutor. For now, the interface is sufficient because you only need to retrieve the ResourceManager to register. Let's assume the ResourceManager runs in the same Actor system as the JobManager until we have an additional leader retrieval for the resource manager.